waspc/data/Generator/templates/react-app/src/operations/resources.js

```diff
@@ -1,12 +1,21 @@
-{{= {= =} =}}
 import { queryClientInitialized } from '../queryClient'
+import { makeCounter } from './counter'
 
 
 // Map where key is resource name and value is Set
 // containing query ids of all the queries that use
 // that resource.
 const resourceToQueryCacheKeys = new Map()
 
+// A counter that counts many actions are currently in progress that will
+// invalidate a given queryCacheKey. It helps us stop premature invalidation and
+// UI flickering after optimistic updates.
+const actionCounter = makeCounter(
+  (queryCacheKey) => {
+    return Array.isArray(queryCacheKey) ? queryCacheKey[0] : queryCacheKey
+  }
+)
```

Comment:
Setting the hash function simply to `hashQueryKey` (react-query's internal function for consistently hashing different query keys) causes excess cache invalidation and, by extension, flickering UI.

This happens because Wasp doesn't know which specific task an action affects (i.e., it doesn't have a notion of what identifies a particular item).

When we receive a response from updating a task with id `1`, we invalidate all queries `get-task` queries (i.e., caches for all tasks, regardless of their IDs). We would ideally want to only invalidate the query with key `['get-task', { id: 1 }` . Unfortunately, Wasp does not enable developers to say something like "This action invalidates a query that fetches a single task **identified by its IDs**,"  so we have no choice but to invalidate all of them.

This is what happens:
1. A user performs an action "set Task 1 to done"
2. The action should optimistically update queries `'operations/get-tasks'` and `['operations/get-task, { id: 1 }']`. We register those two in the counter.
3. The user toggles the task's status one more time before the response arrives (incrementing the counter of both queries to 2). Because we're optimistically updating the UI, it now shows the task as "undone", but two requests have been sent to the server:
    - One request that will mark its status as "done"
    - A second request that will return it to "undone"
4. The first response arrives. We check the counters. Neither of the counters is equal to zero, meaning that there are still actions that will affect the results of these queries in progress. Therefore, we don't invalidate any of queries on the basis of being optimistically updated
5. However, we still want to invalidate queries based on resources declared in the .wasp file, meaning we should invalidate `'operations/get-task'` and `'operations/get-task'`. We check the counters again. `'operations/get-tasks'` still has one related action in progress so we skip it. `'operations/get-task'`doesn't have any actions in progress (because we're keeping count of  `['operations/get-task, { id: 1 }']`. We therefore invalidate `'operations/get-tasks'`, which invalidates all tasks (including `['operations/get-task, { id: 1 }']`, whose counter is still at one).
6. The invalidation of `['operations/get-task, { id: 1 }']` causes a query refetch.
7. The refetch response arrives and changes the UI to show the task's status as "done"
8. The second action response arrives, steps 4-7 are repeated and the UI changes again to show the task's status as "undone"

There are several possible approaches to solving this (e.g., RTK stuff), but I won't list them now.


The "hash" function I used to initialize the counter tries to avoid the flickering by focusing only on the "main" part of the query key. In other words, we are optimistically updating `['operations/get-task, { id: 1 }']`, but the counter will count as if we're updating `'operations/get-task'`. There are two main problems with this approach:
- The function looks pretty fragile/hacky (although, since Wasp controls the query keys, it will technically always work)
- We have now caused a lack of necessary updates because **no task queries will be refresh until all related actions return** (and these actions are only relevant to a single task, not all of them).